tests/workload/transactions.go

```diff
@@ -0,0 +1,125 @@
+	// and corresponding assertions.
+	NewBasicTxWorkload(uri string) (TxWorkloadIterator, error)
+	// Generates a new TxWorkloadIterator that generates a sequence of transactions of the given size.
+	NewSizedTxWorkload(uri string, size int) (TxWorkloadIterator, error)
+}
+
+// TxWorkloadIterator provides an interface for generating a sequence of transactions and corresponding assertions.
+// The caller must proceed in the following sequence:
+// 1. Next
+// 2. GenerateTxWithAssertion
+// 3. (Optional) execute the returned assertion on an arbitrary number of URIs
+type TxWorkloadIterator interface {
+	// Next returns true iff there are more transactions to generate.
+	Next() bool
+	// GenerateTxWithAssertion generates a new transaction and an assertion function that confirms
+	// 1. The tx was accepted on the provided URI
+	// 2. The state was updated as expected according to the provided URI
+	GenerateTxWithAssertion(context.Context) (*chain.Transaction, func(ctx context.Context, uri string) error, error)
+}
+
+func ExecuteWorkload(ctx context.Context, require *require.Assertions, uris []string, generator TxWorkloadIterator) {
+	submitClient := rpc.NewJSONRPCClient(uris[0])
+
+	for generator.Next() {
+		tx, confirm, err := generator.GenerateTxWithAssertion(ctx)
+		require.NoError(err)
+
+		_, err = submitClient.SubmitTx(ctx, tx.Bytes())
+		require.NoError(err)
+
+		utils.ForEach(func(uri string) {
+			err := confirm(ctx, uri)
+			require.NoError(err)
+		}, uris)
+	}
+}
+
+func GenerateNBlocks(ctx context.Context, require *require.Assertions, uris []string, factory TxWorkloadFactory, n uint64) {
+	uri := uris[0]
+	generator, err := factory.NewSizedTxWorkload(uri, int(n))
+	require.NoError(err)
+	client := rpc.NewJSONRPCClient(uri)
+
+	_, startHeight, _, err := client.Accepted(ctx)
+	require.NoError(err)
+	height := startHeight
+	targetheight := startHeight + n
+
+	for generator.Next() && height < targetheight {
+		tx, confirm, err := generator.GenerateTxWithAssertion(ctx)
+		require.NoError(err)
+
+		_, err = client.SubmitTx(ctx, tx.Bytes())
+		require.NoError(err)
+
+		require.NoError(confirm(ctx, uri))
+
+		_, acceptedHeight, _, err := client.Accepted(ctx)
+		require.NoError(err)
+		height = acceptedHeight
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+	utils.ForEach(func(uri string) {
+		client := rpc.NewJSONRPCClient(uri)
+		err := rpc.Wait(ctx, func(ctx context.Context) (bool, error) {
+			_, acceptedHeight, _, err := client.Accepted(ctx)
+			if err != nil {
+				return false, err
+			}
+			return acceptedHeight >= targetheight, nil
+		})
+		require.NoError(err)
+	}, uris)
+}
+
+func GenerateUntilCancel(
+	ctx context.Context,
+	uris []string,
+	generator TxWorkloadIterator,
+) {
+	submitClient := rpc.NewJSONRPCClient(uris[0])
+
+	// Use backgroundCtx within the loop to avoid erroring due to an expected
+	// context cancellation.
+	backgroundCtx := context.Background()
```

Comment:
We can, we just have to do it multiple times and none of these should block. Neither is perfect, but I slightly preferred this to avoid handling the special case error multiple times in the loop because I was not concerned about hanging forever during the actual calls on a cancellation.